source/core/slang-uint-set.h

```diff
@@ -80,16 +141,105 @@ class UIntSet
     static bool hasIntersection(const UIntSet& set1, const UIntSet& set2);
 
-private:
-    enum
+    struct Iterator
     {
-        kElementShift = 5,                              ///< How many bits to shift to get Element index from an index
-        kElementSize = sizeof(Element) * 8,             ///< The number of bits in an element
-        kElementMask = kElementSize - 1,                ///< Mask to get shift from an index
+        friend class UIntSet;
+    private:
+        const List<Element>* context;
+        Index block = 0;
+        Element processedElement = 0;
+        uint64_t LSB = 0;
+
+        void clearLSB()
+        {
+            LSB = bitscanForward(processedElement);
+            processedElement &= processedElement - 1;
+        }
+    public:
+        Iterator(const List<Element>* inContext)
+        {
+            context = inContext;
+        }
+
+        Element operator*()
+        {
+            return Element(LSB + (kElementSize * block));
+        }
+
+        template<typename T>
+        T get()
+        {
+            return T(*(*this));
+        }
+
+        Iterator& operator++()
+        {
+            while (processedElement == 0)
+            {
+                block++;
+                if (block >= context->getCount())
+                {
+                    processedElement = 0;
+                    return *this;
+                }
+                processedElement = (*context)[block];
+            }
+            clearLSB();
+            return *this;
+        }
+        Iterator& operator++(int)
+        {
+            return ++(*this);
+        }
+        bool operator==(const Iterator& other) const
+        {
+            return other.block == this->block
+                && other.processedElement == this->processedElement;
+        }
+        bool operator!=(const Iterator& other) const
+        {
+            return !(other == *this);
+        }
     };
+    Iterator begin() const
+    {
+        Iterator tmp(&m_buffer);
+        if (m_buffer.getCount() == 0)
+            return tmp;
+
+        tmp.processedElement = m_buffer[0];
+        if (tmp.processedElement == 0)
+            tmp++;
+
+        tmp.clearLSB();
+
+        return tmp;
+    }
+    Iterator end() const
+    {
+        Iterator tmp(&m_buffer);
+        tmp.block = m_buffer.getCount();
+        tmp.processedElement = 0;
+        return tmp;
+    }
+
+    bool areAllZero()
+    {
+        return _areAllZero(m_buffer.getBuffer(), m_buffer.getCount());
+    }
 
-    // Make sure they are correct for the Element type
-    SLANG_COMPILE_TIME_ASSERT((1 << kElementShift) == kElementSize);
+protected:
+    static bool _areAllZero(const UIntSet::Element* elems, Index count)
+    {
+        for (Index i = 0; i < count; ++i)
+        {
+            if (elems[i])
+            {
+                return false;
+            }
+        }
+        return true;
+    }
```

Comment:
Probably more efficent to check in a reverse order because the block must have something that causes the size to be increased to the `count`.

Reply:
Capabilities store for stage/target in element [0]/[1], and for extensions we use [1]/[2] normally.

UIntSet uses 'over-allocate' to save on allocation time as well.

Reply:
I see.
But allocation size is different from count, isn't it?

Reply:
no, the count may also be set with leading 0's

Reply:
if a set element becomes 0, we do not remove it